Arguments for fields of interfaces

kl4n4 opened the pull request:
Implements arguments for interfaces (#260 and #261)

allows for GraphQL schemas like the following to be defined:

```graphql
interface Article {
  id: ID!
  headline: String
  image(height: Float!, width: Float!): String!
}

type WebArticle implements Article {
  id: ID!
  headline: String
  image(height: Float!, width: Float!): String!
  url: String!
}
```

using the following syntax:

```typescript
@InterfaceType()
abstract class Article {
    @Field(type => ID)
    id: string;

    @Field({ nullable: true })
    headline?: string;

    @Field()
    image1(@Arg("width") width: number, @Arg("height") height: number): string {
        return `http://lorempixel.com/${width}/${height}/`;
    }
}

@ObjectType({ implements: Article })
class WebArticle extends Article {
    @Field()
    url: string;
}
```

19majkel94 commented:
@kl4n4 
Please rebase your changes, I've made some changes and force pushed on master.

19majkel94 commented on tests/functional/interfaces-with-arguments.ts:
```diff
@@ -0,0 +1,105 @@
+          return `http://lorempixel.com/${width}/${height}/`;
+        }
+        @Field()
+        image2(@Args(type => ImageArgs) args: any): string {
```
I think one sample field method is enough for testing purposes

kl4n4 replied:
I added multiple to test the difference between inline arguments and using an arguments type. And it even turned out to be quite useful in the newly added functional tests that were suggested by you.

19majkel94 replied:
Ok, I haven't spotted that because of not meaningful field names 😉

19majkel94 commented on tests/functional/interfaces-with-arguments.ts:
```diff
@@ -0,0 +1,105 @@
+      class SampleResolver {
+        @Query(returns => [Article])
+        sampleQuery(): Article[] {
+          return [];
```
You should test the resolvers too, not only the signature - does it support resolvers inheritance? Will it call the method from super class correctly?

kl4n4 replied:
you are totally right - and there is an issue with resolving fields of interfaces when using inline arguments. I will have to look deeper into that another day. But I already committed and pushed functional tests for the resolver.

kl4n4 replied:
this fixed - it was missing the field resolvers - I should have used the same arguments as for the object type from the start ;)

19majkel94 commented on tests/functional/interfaces-with-arguments.ts:
```diff
@@ -0,0 +1,105 @@
+        url: string;
+
+        @Field()
+        image3(@Arg("width") width: number, @Arg("height") height: number): string {
```
Please use more meaningfull testing names - `interfaceFieldMethod` and `implemetingObjectTypeFieldMethod`, just like in other test suites

kl4n4 replied:
okay, done :)

19majkel94 commented on README.md:
```diff
@@ -145,7 +145,7 @@ I encourage you to give it a try and experiment with TypeGraphQL. If you have an
 ## Support
 
-TypeGraphQL is an MIT-licensed open source project. This framework is a result of the tremendous amount of work - sleepless nights, busy evenings and weekends. 
+TypeGraphQL is an MIT-licensed open source project. This framework is a result of the tremendous amount of work - sleepless nights, busy evenings and weekends. TypeGraphQL doesn't have a large company that sits behind - its ongoing development is possible only thanks to the support by the community.
```
Still phantom changes 😕

kl4n4 replied:
hmm, I now revert this manually - let me know if this is okay for you

kl4n4 commented:
currently there is an issue with resolving fields of interfaces when using inline arguments. I will investigate and fix this ASAP but I already pushed the tests for reference. As soon as this is fixed I will update the change log and docs.

kl4n4 commented:
Hi @19majkel94, I now fixed all open issues, updated the changelog, added a small doc part and extended the interface example.
I hope now everything looks okay to you - if you still have any issues or suggestions, please let me know. Otherwise I hope this PR helps out one or two other users ;)

19majkel94 commented:
@kl4n4 
Are you going to continue working on this PR? 😉

kl4n4 commented:
sry for the hiatus - my work&projects kept me busy lately but I should find the time now during the easter break.

kl4n4 commented:
@19majkel94 unfortunately it again took me a little longer than expected to finish this. But now it should be done :) I added additional tests for the use case you described in your comments, changed the example back to the way you intended it and adapt everything I could find and think of according to your comments. 

please just let me know, if I missed something or there are still remaining issue with the way I implemented this feature.

19majkel94 commented on tests/functional/interfaces-with-arguments.ts:
```diff
@@ -0,0 +1,406 @@
+          return `${intValue1}-${intValue2}`;
+        }
+        @Field()
+        implemetingObjectTypeFieldArgumentsType(@Args() args: SampleArgs1): string {
```
![image](https://user-images.githubusercontent.com/10618781/56867011-7fd88d80-69e0-11e9-9549-31b386821e4b.png)

19majkel94 commented on tests/functional/interfaces-with-arguments.ts:
```diff
@@ -0,0 +1,406 @@
+      expect(sampleField.args).toBeDefined();
+      expect(sampleField.args.length).toEqual(2);
+      expect(
+        sampleField.args.every(arg => ["intValue1", "intValue2"].includes(arg.name)),
```
Less magic, more explicitness - please use `sampleField.args.find` to get two variables and normal assertion, just like in `resolvers` test suite, instead of `every`, `includes` and `toBeTruthy`.

19majkel94 commented on tests/functional/interfaces-with-arguments.ts:
```diff
@@ -0,0 +1,406 @@
+
+      @ArgsType()
+      class SampleArgs1 {
+        @Field(type => Int, { nullable: true, defaultValue: 50 })
```
Remove the nullable and defaultValue noise if it's not needed or rename the args to reflect that if it's a part of a test case.

19majkel94 replied:
Maybe `inlineIntValue1` and `argsIntValue1` to distinguish the inline from class args in tests.

19majkel94 commented on tests/functional/interfaces-with-arguments.ts:
```diff
@@ -0,0 +1,406 @@
+    });
+
+    it("should have proper arguments for the interface field with inline arguments", async () => {
+      const sampleField = (schemaIntrospection.types.find(
```
Please use temporary variables instead of non-readable long chaining with types assertion. `sampleField` can be a variable initialized in `beforeAll`.

19majkel94 commented on tests/functional/interfaces-with-arguments.ts:
```diff
@@ -0,0 +1,406 @@
+      }
+
+      @ObjectType({ implements: SampleInterface1 })
+      class SampleImplementingObject1 implements SampleInterface1 {
```
The `ObjectType extends InterfaceType` is a nice pattern to reduce the boilerplate of redeclaring fields, we should support that too as it's handy with sharing implementation of interface field resolver. So please test that behavior in both schema and functional way and also check overriding the inline field resolvers methods from base interface.

chrisdostert commented:
is it possible to get this merged? I'm also eagerly waiting on it. Thanks @kl4n4  and @19majkel94 !

19majkel94 commented:
> is it possible to get this merged?

@chrisdostert 
If you have some time you can help and change the requested parts of the code 💪

chrisdostert commented:
@MichalLytek  I hate this answer which is why I didn't respond right away, but I don't have the time to dig into this project to get comfortable enough to contribute : / . I also have FOSS projects outside of my day & night jobs that suck up all my extra time
(https://github.com/opctl and https://github.com/opspec-pkgs to name a couple) . 

Going through issues, multiple have been filed asking for this functionality. Is it possible to get the PR in so people can start using the functionality, and touchups to the implementation can always be revisited?

I would hate to see the amount of work @kl4n4 has gone to go to waste : /

chrisdostert commented:
@MichalLytek @kl4n4 any update on this?

MichalLytek commented:
Due to lack of ["Allowing changes to a pull request branch created from a fork"](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/), I am forced to close this PR in favor of the revamped #579 🔒

MichalLytek commented:
#579 has been merged! 🎉 

Thank you very much @kl4n4 for your work on this PR ❤️ 
It's a bit said that you haven't find time to finish this PR and a bit shame for me that it took one year to finish this simple feature 😞
